stl/inc/deque

```diff
@@ -1557,9 +1557,15 @@ private:
             _Newsize *= 2;
         }
 
+        size_type _Allocsize = _Newsize;
+
         size_type _Myboff = _Myoff() / _Block_size;
-        _Mapptr _Newmap   = _Allocate_at_least_helper(_Almap, _Newsize);
+        _Mapptr _Newmap   = _Allocate_at_least_helper(_Almap, _Allocsize);
         _Mapptr _Myptr    = _Newmap + _Myboff;
+        _STL_ASSERT(_Allocsize >= _Newsize, "_Allocsize >= _Newsize");
+        while (_Allocsize / 2 >= _Newsize) {
+            _Newsize *= 2;
+        }

```

Comment:
This seems inefficient to try and get the biggest power of two less than `_Allocsize`.

Reply:
Do we have an `_Ugly` (i.e., usable in C++14 mode) version of `bit_floor`?

Reply:
I think it's ok to (temporarily) keep this while loop, as it's highly likely that `_Allocsize` is just equal to or marginally larger than `_Newsize` so the loop will skip immediatly or very short (And I'm going to do a series of cleanups for `<deque>` recently :) we can consider making its above while loop to a bit_ceil too )

Reply:
I made a tiny cleanup & optimization for its above loop: https://github.com/microsoft/STL/commit/850d8aee1e8abadfff47f18726be6897c94dc5c4; I think this can partly compensate for the cost of new loop. Can I make a push to this pr?

Reply:
[godbolt result](https://godbolt.org/z/E5fbM4Ysn); the codegen looks really better

Reply:
hmm, I'm having difficulty deciding whether we can safely assume `_Minimum_map_size` will not cause overflow against `max_size()`.
The optimization may become "buggy" under some extreme (non practical) cases (for example, a `deque<char>` with an allocator whose `max_size()` is less than `128`.

And worse, the comparision against `_Alty::max_size` in current code looks wrong to me. I'm doubting we should use `_Almap::max_size` instead.

Due to these problems I'd suggest putting off that commit to future prs. (And in that link I've added some comments about the problems in detail.)